File: Data Structures & Algorithms/is-anagram/submission-0.cpp
class Solution {
public:
    bool isAnagram(string s, string t) {

        int n = s.length();

        if (n != t.length()) {
            return false;
        }

        unordered_map<char, int> sMap;

        for (int i = 0; i < n; i++) {
            if (sMap.find(s[i]) != sMap.end()) {
                sMap[s[i]]++;
            }
            else {
                sMap[s[i]] = 1;
            }
        }

        for (int j = 0; j < n; j++) {
            if (sMap.find(t[j]) != sMap.end()) {
                if (sMap[t[j]] > 0) {
                    sMap[t[j]]--;
                }
                else {
                    return false;
                }
            }
            else {
                return false;
            }
        }

        return true;


    }
};

File: Data Structures & Algorithms/count-connected-components/submission-0.py
class DSU:
    def __init__(self, n):
        self.parent = [i for i in range(n)]
        self.rank = [1] * n
        self.components = n
    
    def find(self, u):
        if self.parent[u] != u:
            self.parent[u] = self.find(self.parent[u])
        return self.parent[u]
    
    def union(self, u, v):
        u = self.find(u)
        v = self.find(v)

        if u == v:
            return
        
        if self.rank[u] < self.rank[v]:
            self.parent[u] = v
            self.rank[v] += self.rank[u]
        else:
            self.parent[v] = u
            self.rank[u] += self.rank[v]
        
        self.components -= 1
        

class Solution:
    def countComponents(self, n: int, edges: List[List[int]]) -> int:
        dsu = DSU(n)

        for u, v in edges:
            dsu.union(u, v)
        
        return dsu.components
        